src/networking.c

```diff
@@ -3559,564 +3620,681 @@ void quitCommand(client *c) {
-            addReplyArrayLen(c, 0);
+static void clientCommandNoTouch(client *c) {
+    /* CLIENT NO-TOUCH ON|OFF */
+    if (!strcasecmp(c->argv[2]->ptr, "on")) {
+        c->flag.no_touch = 1;
+        addReply(c, shared.ok);
+    } else if (!strcasecmp(c->argv[2]->ptr, "off")) {
+        c->flag.no_touch = 0;
+        addReply(c, shared.ok);
+    } else {
+        addReplyErrorObject(c, shared.syntaxerr);
+    }
+}
+
+static void clientCommandCapa(client *c) {
+    for (int i = 2; i < c->argc; i++) {
+        if (!strcasecmp(c->argv[i]->ptr, "redirect")) {
+            c->capa |= CLIENT_CAPA_REDIRECT;
         }
+    }
+    addReply(c, shared.ok);
+}
+
+static void clientCommandImportSource(client *c) {
+    /* CLIENT IMPORT-SOURCE ON|OFF */
+    if (!server.import_mode && strcasecmp(c->argv[2]->ptr, "off")) {
+        addReplyError(c, "Server is not in import mode");
+        return;
+    }
+    if (!strcasecmp(c->argv[2]->ptr, "on")) {
+        c->flag.import_source = 1;
+        addReply(c, shared.ok);
+    } else if (!strcasecmp(c->argv[2]->ptr, "off")) {
+        c->flag.import_source = 0;
+        addReply(c, shared.ok);
+    } else {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+}
+
+void clientCommand(client *c) {
+    if (c->argc == 2 && !strcasecmp(c->argv[1]->ptr, "help")) {
+        clientCommandHelp(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "id") && c->argc == 2) {
+        clientCommandID(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "info") && c->argc == 2) {
+        clientCommandInfo(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "list")) {
+        clientCommandList(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "reply") && c->argc == 3) {
+        clientCommandReply(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "no-evict") && c->argc == 3) {
+        clientCommandNoEvict(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "kill")) {
+        clientCommandKill(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "unblock") && (c->argc == 3 || c->argc == 4)) {
+        clientCommandUnblock(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "setname") && c->argc == 3) {
+        clientCommandSetName(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "getname") && c->argc == 2) {
+        clientCommandGetName(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "unpause") && c->argc == 2) {
+        clientCommandUnpause(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "pause") && (c->argc == 3 || c->argc == 4)) {
+        clientCommandPause(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "tracking") && c->argc >= 3) {
+        clientCommandTracking(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "caching") && c->argc >= 3) {
+        clientCommandCaching(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "getredir") && c->argc == 2) {
+        clientCommandGetredir(c);
+    } else if (!strcasecmp(c->argv[1]->ptr, "trackinginfo") && c->argc == 2) {
+        clientCommandTrackingInfo(c);
```

Comment:
Do we need this entire piece of refactoring though? It would be ideal if we can limit the changes to the scope of the PR.

Reply:
@hpatro I actually suggested to split it into smaller functions in this comment: https://github.com/valkey-io/valkey/pull/1401#discussion_r1880090690

For CLIENT LIST and CLIENT KILL, it makes the code simpler. Technically, we don't need to do it for all the others though if we want to minimize the diff. This refactoring was very cleanly done though, so the body of each subcommand implementation doesn't appear to be changed in the diff. They're still in the same order and unchanged.

Reply:
It might sound counterintuitive to modern best practices but I got used to not jumping around the IDE with this project and look at these if blocks together in one place. I don't have a strong argument though.

Reply:
Yeah, I agree, and we usually want to avoid unnecessary changes.

What I commented on before this refactoring was that in an earlier commit in this PR, there were multiple `zfree` on each error case, something like this:

```C
        if (...) {
            if (parseClientFilters(c, i, &client_filter) != C_OK) {
                zfree(client_filter.ids);
                return;
            }
        } else {
            zfree(client_filter.ids);
            addReplyErrorObject(c, shared.syntaxerr);
            return;
        }

        /* finally, the non-error case */
        ...
        zfree(client_filter.ids);
```

There were different ways to solve it, not necessarily split it up into multple functions. This was solved first by a goto, like this:

```C
        if (...) {
            if (parseClientFilters(c, i, &client_filter) != C_OK) {
                goto client_kill_done;
            }
        } else {
            addReplyErrorObject(c, shared.syntaxerr);
            goto client_kill_done;
        }

        /* finally, the non-error case */
        ...

client_kill_done:
        zfree(client_filter.ids);
```

Perhaps the `client_kill_done` label in the middle of a large function seemed odd and that's a reason to break CLIENT KILL into a separate function? I don't know, but I didn't want to be too picky about exactly how a contributor formulates the code. :)

@hpatro if you prefer to avoid the large chain of else-if, we could revert it, or there is another way: We could change `"function": "clientCommand"` in `src/commands/client-kill.json` to point directly to the function implementing this subcommand, `"function": "clientKillCommand"`, and similar. It would have to take the argv and argc arguments.

What's best?